### ISSUE TITLE
Add a container ID to tools

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bcc#1119]
This is a placeholder to discuss this need.

I think there's three different ways a tool could have such an ID:

- **Match on the namespace**. Not that hard to dig out (at least, in an unstable way by walking from the task_struct). We already match the network namespace from a couple of tools.
- **Match on the cgroup**. Except cgroups are currently being rewritten with cgroup v2. And even if we did cgroup v1, what would we match on? The name in /proc/PID/cgroup is hard to construct, and some people have used the inode number instead (like is shown in /proc/PID/ns).
- **Match on a container ID**. Except that doesn't exist yet in the kernel. Some people have suggested using the cgroup inode.

I added a --pidnss equivalent to runqlat to breakdown by PID namespaces (and it should have a --pidns option too, to match a single namespace), but I'm not yet certain what the final outcome for all tools should be.

In the meantime, I'd likely add namespace matching to some tools, but I wouldn't rush to do it everywhere yet.

**Comment by goldshtn** [comment]
I often get asked about tracing containers from the host system, and I think the general consensus is that it would be very nice to "easily" specify which containers you care about. For some tools, printing the "container ID" or aggregating by it (like you did in runqlat) probably makes sense, but I'd say filtering to a specific container is even more commonly requested.

**Comment by brendangregg** [comment]
Yes, but, what is a container? There is no container ID (yet, anyway) in the kernel. It's namespaces + cgroups (and maybe seccomp, apparmor). Docker gives them container IDs, but that's user space, and Docker specific. There was an idea to use the cgroup inode number (I think), but that's not official.

Let's say it did, and we had picked some kernel ID to also be the container ID. We could then add BPF_func_get_current_CID(), and add this option to all tools easily. But we aren't there yet

**Comment by goldshtn** [comment]
Yes, there's a reason why I put "container ID" in quotes :smile:

**Comment by goldshtn** [comment]
But seriously though, I think it would be helpful for many people if we at least work well with Docker's notion of a container.

**Comment by brendangregg** [comment]
Ok, well, we could write a bpf_get_current_cid() macro that actually called a little BPF program, which did bpf_get_current_task() then picked out the, say, PID namespace inode number (like I did), and then mapped that to the Docker container ID in user space.

Or, to be more explicit, we could have a bpf_get_current_pidns() that did the same. And some Docker helpers:

- pidns_to_dockerid()
- dockerid_to_pidns()

Then the developer knows exactly what's going on. And we could add helpers for other container technologies, if/when need be.

**Comment by goldshtn** [comment]
Yes, that sounds like a great plan.

**Comment by 4ast** [comment]
@brendangregg have you considered using bpf_current_task_under_cgroup() ?
i think that's the cleanest way to match given container.
ns.inum trick that we do in several tools is frowned upon by kernel namespace maintainer,
since he insists that ns.inum is only valid within given /proc filesystem,
hence iproute2/ip/ipnetns.c does
```
      if ((st.st_dev == netst.st_dev) &&
          (st.st_ino == netst.st_ino)) {
```
Today the matching on devno is not necessary, since ns.inum allocation is global,
but we shouldn't be relying on that implementation detail.
So if we can avoid comparing inum that's the best.

**Comment by bobrik** [comment]
`bpf_current_task_under_cgroup()` requires you to pass cgroup in advance. My use case is classifying events similarly to the way cAdvisor classifies cgroup metrics:

* systemd units get `id="/system.slice/systemd-journald.service"`
* docker containers get `id="/docker/7d52efc8bb5a1f25b377..."`

Both IDs are CPU or memory cgroup names. It's then trivial to map both systemd units (which are not in any namespace) and docker containers to something usable in userspace where you can query both in any way you like.

cc @sargun for ideas.

**Comment by yonghong-song** [comment]
I guess the complex will be mostly in user space. To require user find and provide the lengthy string id like `id="/system.slice/systemd-journald.service"` is not really ideal. What are better ways we could do? For example, user could specify a `pid` or something, then the script automatically derives the cgroup (the deepest nested one)? What are common use cases?

**Comment by bobrik** [comment]
My use-case is to track ALL processes and then map them to containers or systemd units in user space.

See https://github.com/cloudflare/ebpf_exporter/issues/13.

**Comment by bobrik** [comment]
I've had some progress with this.

With shared maps I can run one program to map PID -> cgroup path. Comment describing workflow:

```
// Possible events (root id=5 means cpu cgroup, from /proc/cgroups):
// * CGroup creation:
//   * cgroup:cgroup_mkdir root=5 id=102 level=2 path=/system.slice/mtail.service
// * Adding task to cgroup:
//   * cgroup:cgroup_attach_task: dst_root=5 dst_id=102 dst_level=2 dst_path=/system.slice/mtail.service pid=13004 comm=(apper.sh)
// * Removing cgroup (no task detachment!):
//   * cgroup:cgroup_rmdir: root=5 id=43 level=2 path=/system.slice/mtail.service
//
// We need the following maps then (only for one specific dst_root):
// * Map from pid to cgroup id
// * Map from cgroup id to cgroup path
//
// Workflow is the following:
// * On "cgroup:cgroup_attach_task" and "cgroup:cgroup_transfer_tasks" we map pid -> dst_path
// * On "syscalls:sys_enter_exit" and "syscalls:sys_enter_exit_group" we remove pid from the map
// * After setting up tracepoints we iterate sysfs and populate map
//
// PIDs do not wrap around often, so this should not cause any visible races.
```

This map then can be used by another eBPF program to produce a map with cgroup path as the key.

Userspace then can map cgroup name to whatever it wants, which usually is systemd unit or docker container name / tag / label.

This is all assuming that "containers" have some cgroup assigned (works for sytemd and docker).

Does this sound like it could be a reasonable solution?

**Comment by yonghong-song** [comment]
There is a kernel RFC https://patchwork.kernel.org/patch/10396321/ to get the cgroup inode number for the `current` task during tracing. This could be helpful in this case. 

Basically, the program can record both `pid` and `cgroup_inode` at the same time. In user space, the cgroup hierarchy can be traversed to match `cgroup_inode` and this can be built in advance. During printout phase, the python program can just simply print out something like `container_id[cgroup_inode]`. pid's may come and go pretty fast. cgroup hierarchy should be relatively stable.

**Comment by bobrik** [comment]
@yonghong-song I'm not very familiar with cgroup internals, is `cgroup_inode` the same as `id` in tracepoints? Is there a way to reliably build and update `cgroup_inode` -> `cgroup_path` without tracepoints in userspace?

The idea overall sounds good to me and the patch looks small enough to patch our LTS.

cc @alban for usecase discussion
